models/user_openid.go

```diff
@@ -0,0 +1,152 @@
+	"fmt"
+	"strings"
+
+	"code.gitea.io/gitea/modules/auth/openid"
+)
+
+var (
+	// ErrOpenIDNotExist openid is not known
+	ErrOpenIDNotExist = errors.New("OpenID is unknown")
+)
+
+// UserOpenID is the list of all OpenID identities of a user.
+type UserOpenID struct {
+	ID          int64  `xorm:"pk autoincr"`
+	UID         int64  `xorm:"INDEX NOT NULL"`
+	URI         string `xorm:"UNIQUE NOT NULL"`
+}
+
+// GetUserOpenIDs returns all openid addresses that belongs to given user.
+func GetUserOpenIDs(uid int64) ([]*UserOpenID, error) {
+	openids := make([]*UserOpenID, 0, 5)
+	if err := x.
+		Where("uid=?", uid).
+		Find(&openids); err != nil {
+		return nil, err
+	}
+
+	return openids, nil
+}
+
+func isOpenIDUsed(e Engine, uri string) (bool, error) {
+	if len(uri) == 0 {
+		return true, nil
+	}
+
+	return e.Get(&UserOpenID{URI: uri})
+}
+
+// IsOpenIDUsed returns true if the openid has been used.
+func IsOpenIDUsed(openid string) (bool, error) {
+	return isOpenIDUsed(x, openid)
+}
+
+func addUserOpenID(e Engine, openid *UserOpenID) error {
+	openid.URI = strings.ToLower(strings.TrimSpace(openid.URI))
+	used, err := isOpenIDUsed(e, openid.URI)
+	if err != nil {
+		return err
+	} else if used {
+		return ErrOpenIDAlreadyUsed{openid.URI}
+	}
+
+	_, err = e.Insert(openid)
+	return err
+}
+
+// AddUserOpenID adds an openid address to given user.
+func AddUserOpenID(openid *UserOpenID) error {
+	return addUserOpenID(x, openid)
+}
+
+// AddUserOpenIDs adds an openid address to given user.
+func AddUserOpenIDs(openids []*UserOpenID) error {
+	if len(openids) == 0 {
+		return nil
+	}
+
+	// Check if any of them has been used
+	for i := range openids {
+		// TODO: normalize ?
+		openids[i].URI = strings.ToLower(strings.TrimSpace(openids[i].URI))
+		used, err := IsOpenIDUsed(openids[i].URI)
+		if err != nil {
+			return err
+		} else if used {
+			return ErrOpenIDAlreadyUsed{openids[i].URI}
+		}
+	}
+
+	if _, err := x.Insert(openids); err != nil {
+		return fmt.Errorf("Insert: %v", err)
+	}
+
+	return nil
+}
+
+// DeleteUserOpenID deletes an openid address of given user.
+func DeleteUserOpenID(openid *UserOpenID) (err error) {
+	var deleted int64
+	// ask to check UID
+	var address = UserOpenID{
+		UID: openid.UID,
+	}
+	if openid.ID > 0 {
+		deleted, err = x.Id(openid.ID).Delete(&address)
+	} else {
+		deleted, err = x.
+			Where("openid=?", openid.URI).
+			Delete(&address)
+	}
+
+	if err != nil {
+		return err
+	} else if deleted != 1 {
+		return ErrOpenIDNotExist
+	}
+	return nil
+}
+
+// DeleteUserOpenIDs deletes multiple openid addresses
+func DeleteUserOpenIDs(openids []*UserOpenID) (err error) {
+	for i := range openids {
+		if err = DeleteUserOpenID(openids[i]); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// GetUserByOpenID returns the user object by given OpenID if exists.
+func GetUserByOpenID(uri string) (*User, error) {
+	if len(uri) == 0 {
+		return nil, ErrUserNotExist{0, uri, 0}
+	}
+
+	uri, err := openid.Normalize(uri)
```

Comment:
Fixed with 2ba2d6a7303f1e1cbc26dbc2f5aad938f4afa37e (the addUser function assumes incoming item is pre-normalized as it returns from a verification step)